File: fpga/phase1_ethernet/ps_gateway/gateway.py
#!/usr/bin/env python3
"""
Eclypse Z7 PS Phase-1 gateway (Python).

- TCP control server on 7600
- UDP TEST_COUNTER stream to MetaField host while capturing

Run on the board:
  export MF_HOST=192.168.1.10
  python3 gateway.py
"""

from __future__ import annotations

import os
import socket
import struct
import threading
import time

MAGIC = 0x4D464530  # MFE0
VERSION = 1
HEADER_FMT = "<IHH IQ H H I I H H"
MSG_TEST_COUNTER = 0x0001

HOST = os.environ.get("MF_HOST", "192.168.1.10")
DATA_PORT = int(os.environ.get("MF_DATA_PORT", "7601"))
CTRL_PORT = int(os.environ.get("MF_CTRL_PORT", "7600"))
RATE_HZ = float(os.environ.get("MF_RATE_HZ", "1000"))

_state = {
    "capturing": False,
    "sample_rate": 1_000_000,
    "channels": [0, 1, 2, 3],
    "mode": "RAW",
    "sequence": 0,
    "lock": threading.Lock(),
}


def monotonic_ticks() -> int:
    # Placeholder acquisition clock — replace with PL timer mmap later
    return time.monotonic_ns()


def pack_test_counter(sequence: int, timestamp_ticks: int) -> bytes:
    hdr = struct.pack(
        HEADER_FMT,
        MAGIC,
        VERSION,
        MSG_TEST_COUNTER,
        sequence & 0xFFFFFFFF,
        timestamp_ticks & 0xFFFFFFFFFFFFFFFF,
        0,  # source_id (PS test)
        0,  # channel
        int(_state["sample_rate"]) & 0xFFFFFFFF,
        1,  # sample_count
        1,  # payload_type INT16
        0,  # flags
    )
    body = struct.pack("<QII", timestamp_ticks, sequence & 0xFFFFFFFF, 0)
    return hdr + body


def handle_line(line: str) -> str:
    parts = line.strip().split()
    if not parts:
        return "ERR empty"
    cmd = parts[0].upper()

    with _state["lock"]:
        if cmd == "GET_STATUS":
            return (
                f"OK capturing={int(_state['capturing'])} "
                f"seq={_state['sequence']} "
                f"rate={_state['sample_rate']} "
                f"mode={_state['mode']} "
                f"channels={','.join(map(str, _state['channels']))}"
            )
        if cmd == "START_CAPTURE":
            _state["capturing"] = True
            return "OK START_CAPTURE"
        if cmd == "STOP_CAPTURE":
            _state["capturing"] = False
            return "OK STOP_CAPTURE"
        if cmd == "SET_SAMPLE_RATE" and len(parts) >= 2:
            _state["sample_rate"] = int(float(parts[1]))
            return f"OK sample_rate={_state['sample_rate']}"
        if cmd == "SET_CHANNELS" and len(parts) >= 2:
            _state["channels"] = [int(x) for x in parts[1].split(",") if x != ""]
            return f"OK channels={_state['channels']}"
        if cmd == "SET_MODE" and len(parts) >= 2:
            _state["mode"] = parts[1].upper()
            return f"OK mode={_state['mode']}"
        if cmd == "PING":
            return "OK PONG"

    return f"ERR unknown command: {cmd}"


def control_server() -> None:
    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    srv.bind(("0.0.0.0", CTRL_PORT))
    srv.listen(4)
    print(f"[ctrl] TCP listening on 0.0.0.0:{CTRL_PORT}")

    while True:
        conn, addr = srv.accept()
        with conn:
            data = b""
            while True:
                chunk = conn.recv(1024)
                if not chunk:
                    break
                data += chunk
                if b"\n" in data:
                    break
            line = data.decode("utf-8", errors="replace").splitlines()[0] if data else ""
            reply = handle_line(line) + "\n"
            conn.sendall(reply.encode("utf-8"))
            print(f"[ctrl] {addr}: {line!r} -> {reply.strip()}")


def data_emitter() -> None:
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    period = 1.0 / max(RATE_HZ, 1.0)
    print(f"[data] UDP -> {HOST}:{DATA_PORT} at {RATE_HZ} Hz (when capturing)")

    next_t = time.monotonic()
    while True:
        with _state["lock"]:
            capturing = _state["capturing"]
            if capturing:
                seq = _state["sequence"]
                _state["sequence"] = (seq + 1) & 0xFFFFFFFF
            else:
                seq = None

        if seq is not None:
            pkt = pack_test_counter(seq, monotonic_ticks())
            try:
                sock.sendto(pkt, (HOST, DATA_PORT))
            except OSError as e:
                print(f"[data] send error: {e}")

        next_t += period
        sleep = next_t - time.monotonic()
        if sleep > 0:
            time.sleep(sleep)
        else:
            next_t = time.monotonic()


def main() -> None:
    print("MetaField Eclypse PS gateway (Phase 1 Python)")
    print(f"  host={HOST} data_port={DATA_PORT} ctrl_port={CTRL_PORT}")
    threading.Thread(target=control_server, daemon=True).start()
    data_emitter()  # main thread


if __name__ == "__main__":
    main()

File: fpga/phase1_ethernet/ps_gateway/gateway.c
/*
 * Eclypse Z7 PS Phase-1 gateway (C, Linux userspace)
 *
 * Build on board or cross-compile:
 *   gcc -O2 -Wall -o gateway gateway.c
 *
 *   MF_HOST=192.168.1.10 ./gateway
 *
 * TCP :7600 control  |  UDP TEST_COUNTER to host:7601 while capturing
 */

#define _GNU_SOURCE
#include <arpa/inet.h>
#include <errno.h>
#include <netinet/in.h>
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <time.h>
#include <unistd.h>

#define MAGIC 0x4D464530u /* MFE0 */
#define VERSION 1
#define MSG_TEST_COUNTER 0x0001

static char g_host[64] = "192.168.1.10";
static int g_data_port = 7601;
static int g_ctrl_port = 7600;
static double g_rate_hz = 1000.0;

static pthread_mutex_t g_mu = PTHREAD_MUTEX_INITIALIZER;
static int g_capturing = 0;
static uint32_t g_sequence = 0;
static uint32_t g_sample_rate = 1000000;
static char g_mode[32] = "RAW";

static uint64_t monotonic_ticks(void) {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return (uint64_t)ts.tv_sec * 1000000000ull + (uint64_t)ts.tv_nsec;
}

/* Pack little-endian helpers */
static void wr_u16(uint8_t *p, uint16_t v) {
  p[0] = (uint8_t)(v);
  p[1] = (uint8_t)(v >> 8);
}
static void wr_u32(uint8_t *p, uint32_t v) {
  p[0] = (uint8_t)(v);
  p[1] = (uint8_t)(v >> 8);
  p[2] = (uint8_t)(v >> 16);
  p[3] = (uint8_t)(v >> 24);
}
static void wr_u64(uint8_t *p, uint64_t v) {
  for (int i = 0; i < 8; i++)
    p[i] = (uint8_t)(v >> (8 * i));
}

static size_t pack_test_counter(uint8_t *out, uint32_t seq, uint64_t ts) {
  /* 36-byte header + 16-byte body */
  wr_u32(out + 0, MAGIC);
  wr_u16(out + 4, VERSION);
  wr_u16(out + 6, MSG_TEST_COUNTER);
  wr_u32(out + 8, seq);
  wr_u64(out + 12, ts);
  wr_u16(out + 20, 0); /* source */
  wr_u16(out + 22, 0); /* channel */
  wr_u32(out + 24, g_sample_rate);
  wr_u32(out + 28, 1);
  wr_u16(out + 32, 1);
  wr_u16(out + 34, 0);
  wr_u64(out + 36, ts);
  wr_u32(out + 44, seq);
  wr_u32(out + 48, 0);
  return 52;
}

static void handle_line(const char *line, char *reply, size_t reply_len) {
  char cmd[64];
  char arg[128];
  cmd[0] = arg[0] = 0;
  sscanf(line, "%63s %127[^
]", cmd, arg);

  pthread_mutex_lock(&g_mu);
  if (strcmp(cmd, "GET_STATUS") == 0) {
    snprintf(reply, reply_len,
             "OK capturing=%d seq=%u rate=%u mode=%s",
             g_capturing, g_sequence, g_sample_rate, g_mode);
  } else if (strcmp(cmd, "START_CAPTURE") == 0) {
    g_capturing = 1;
    snprintf(reply, reply_len, "OK START_CAPTURE");
  } else if (strcmp(cmd, "STOP_CAPTURE") == 0) {
    g_capturing = 0;
    snprintf(reply, reply_len, "OK STOP_CAPTURE");
  } else if (strcmp(cmd, "SET_SAMPLE_RATE") == 0) {
    g_sample_rate = (uint32_t)strtoul(arg, NULL, 10);
    snprintf(reply, reply_len, "OK sample_rate=%u", g_sample_rate);
  } else if (strcmp(cmd, "SET_MODE") == 0) {
    snprintf(g_mode, sizeof(g_mode), "%s", arg);
    snprintf(reply, reply_len, "OK mode=%s", g_mode);
  } else if (strcmp(cmd, "PING") == 0) {
    snprintf(reply, reply_len, "OK PONG");
  } else {
    snprintf(reply, reply_len, "ERR unknown command: %s", cmd);
  }
  pthread_mutex_unlock(&g_mu);
}

static void *control_thread(void *arg) {
  (void)arg;
  int srv = socket(AF_INET, SOCK_STREAM, 0);
  int yes = 1;
  setsockopt(srv, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));

  struct sockaddr_in addr;
  memset(&addr, 0, sizeof(addr));
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = INADDR_ANY;
  addr.sin_port = htons((uint16_t)g_ctrl_port);

  if (bind(srv, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
    perror("bind ctrl");
    return NULL;
  }
  listen(srv, 4);
  fprintf(stderr, "[ctrl] TCP 0.0.0.0:%d\n", g_ctrl_port);

  for (;;) {
    struct sockaddr_in cli;
    socklen_t clilen = sizeof(cli);
    int c = accept(srv, (struct sockaddr *)&cli, &clilen);
    if (c < 0)
      continue;

    char buf[256];
    ssize_t n = recv(c, buf, sizeof(buf) - 1, 0);
    if (n > 0) {
      buf[n] = 0;
      /* first line only */
      char *nl = strchr(buf, '\n');
      if (nl)
        *nl = 0;
      char reply[256];
      handle_line(buf, reply, sizeof(reply));
      size_t len = strlen(reply);
      reply[len] = '\n';
      reply[len + 1] = 0;
      send(c, reply, len + 1, 0);
      fprintf(stderr, "[ctrl] %s -> %s", buf, reply);
    }
    close(c);
  }
  return NULL;
}

static void data_loop(void) {
  int sock = socket(AF_INET, SOCK_DGRAM, 0);
  struct sockaddr_in dst;
  memset(&dst, 0, sizeof(dst));
  dst.sin_family = AF_INET;
  dst.sin_port = htons((uint16_t)g_data_port);
  inet_pton(AF_INET, g_host, &dst.sin_addr);

  fprintf(stderr, "[data] UDP -> %s:%d @ %.1f Hz\n", g_host, g_data_port, g_rate_hz);

  const double period = 1.0 / (g_rate_hz > 1.0 ? g_rate_hz : 1.0);
  struct timespec next;
  clock_gettime(CLOCK_MONOTONIC, &next);

  uint8_t pkt[64];
  for (;;) {
    int capturing;
    uint32_t seq = 0;

    pthread_mutex_lock(&g_mu);
    capturing = g_capturing;
    if (capturing) {
      seq = g_sequence++;
    }
    pthread_mutex_unlock(&g_mu);

    if (capturing) {
      size_t n = pack_test_counter(pkt, seq, monotonic_ticks());
      sendto(sock, pkt, n, 0, (struct sockaddr *)&dst, sizeof(dst));
    }

    next.tv_nsec += (long)(period * 1e9);
    while (next.tv_nsec >= 1000000000L) {
      next.tv_nsec -= 1000000000L;
      next.tv_sec++;
    }
    clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &next, NULL);
  }
}

int main(int argc, char **argv) {
  const char *env;
  if ((env = getenv("MF_HOST")))
    snprintf(g_host, sizeof(g_host), "%s", env);
  if ((env = getenv("MF_DATA_PORT")))
    g_data_port = atoi(env);
  if ((env = getenv("MF_CTRL_PORT")))
    g_ctrl_port = atoi(env);
  if ((env = getenv("MF_RATE_HZ")))
    g_rate_hz = atof(env);
  if (argc >= 2)
    snprintf(g_host, sizeof(g_host), "%s", argv[1]);

  fprintf(stderr, "MetaField Eclypse PS gateway (Phase 1 C)\n");

  pthread_t tid;
  pthread_create(&tid, NULL, control_thread, NULL);
  pthread_detach(tid);

  data_loop();
  return 0;
}
